File: surface/widgets/nativescript/widgets/primitives/ScrollLock.ts
// ============================================================
// Clef Surface NativeScript Widget — ScrollLock
//
// NativeScript scroll lock container that can enable or
// disable scrolling of its content. Useful for preventing
// background scroll when modals or overlays are open.
// ============================================================

import { ScrollView, StackLayout } from '@nativescript/core';

// --------------- Props ---------------

export interface ScrollLockProps {
  locked?: boolean;
  orientation?: 'vertical' | 'horizontal';
  padding?: number;
}

// --------------- Component ---------------

export function createScrollLock(props: ScrollLockProps = {}): ScrollView {
  const {
    locked = false,
    orientation = 'vertical',
    padding = 0,
  } = props;

  const scrollView = new ScrollView();
  scrollView.className = 'clef-scroll-lock';
  scrollView.orientation = orientation;
  scrollView.isScrollEnabled = !locked;

  const content = new StackLayout();
  content.className = 'clef-scroll-lock__content';
  content.padding = padding;

  if (orientation === 'horizontal') {
    content.orientation = 'horizontal';
  }

  scrollView.content = content;

  // Expose lock/unlock helpers
  const lock = () => {
    scrollView.isScrollEnabled = false;
    scrollView.className = 'clef-scroll-lock clef-scroll-lock--locked';
  };

  const unlock = () => {
    scrollView.isScrollEnabled = true;
    scrollView.className = 'clef-scroll-lock';
  };

  const toggle = () => {
    if (scrollView.isScrollEnabled) lock();
    else unlock();
  };

  (scrollView as any).__clefScrollLock = { lock, unlock, toggle, content };

  return scrollView;
}

createScrollLock.displayName = 'ScrollLock';
export default createScrollLock;

File: surface/widgets/appkit/components/widgets/primitives/ScrollLock.swift
// ============================================================
// Clef Surface AppKit Widget — ScrollLock
//
// Prevents or restores scrolling on a parent NSScrollView.
// Useful when modals or overlays are active to prevent
// background content from scrolling.
// ============================================================

import AppKit

public class ClefScrollLock {
    private weak var scrollView: NSScrollView?
    private var savedHasVertical: Bool = true
    private var savedHasHorizontal: Bool = true

    public var locked: Bool = false {
        didSet { locked ? lock() : unlock() }
    }

    public init(scrollView: NSScrollView? = nil) {
        self.scrollView = scrollView
    }

    public func attach(to scrollView: NSScrollView) {
        self.scrollView = scrollView
    }

    private func lock() {
        guard let sv = scrollView else { return }
        savedHasVertical = sv.hasVerticalScroller
        savedHasHorizontal = sv.hasHorizontalScroller
        sv.hasVerticalScroller = false
        sv.hasHorizontalScroller = false
    }

    private func unlock() {
        guard let sv = scrollView else { return }
        sv.hasVerticalScroller = savedHasVertical
        sv.hasHorizontalScroller = savedHasHorizontal
    }
}
